### PR TITLE
StorageTx improvements

### PR DESCRIPTION
**Pull request by cstamas**
Changes:
- ability to save blobs ahead of touching any Orient entity (w/ cleanup support in TX scope)
- filling in proper blob headers, like txPrincipal, name, etc, no need in various facets to have empty maps created
- moved content validation to StorageTx (along with strictContentTypeValidation config)
- open for custom content validation for format specific reqs
- cleanup of some changes brought in by https://github.com/sonatype/nexus-oss/pull/1327
- Maven2 specific file content validator (https://issues.sonatype.org/browse/NEXUS-8667)
- UI change to migrate strictContentTypeValidation from various facets to StorageFacet

CI
http://bamboo.s/browse/NX3-OSSF330

### CODE REVIEW

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/AssetBlob.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+   * Returns {@code true} if this instance is attached to an {@link Asset}. If {@code false} returned, the blob
+   * referenced by this instance is considered "orphan" and will be deleted at the end of TX (whatever outcome is,
+   * commit or rollback) if not attached until the end of TX.
+   */
+  boolean isAttached() {
+    return attached;
+  }
+
+  /**
+   * Set's the attached state or this instance. Only can be invoked once, while this instance is not attached.
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/AssetBlob.java]: Sets

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/AssetBlob.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Blob handle, that holds properties of newly created Blob that is about to be assigned to an {@link Asset}.
+ *
+ * @since 3.0
+ */
+public class AssetBlob
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/AssetBlob.java]: I wonder if BlobHandle is a better name, especially since orphan 'AssetBlobs' don't have assets.

Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/AssetBlob.java]: We had discussion about it's name already, Stuart proposed AssetBlob, as it MUST be attached to an Asset, otherwise it's cleaned up as orphan.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+      if (contentType == null && strictContentTypeValidation) {
+        throw new InvalidContentException("Content type could not be determined.");
+      }
+    }
+    else {
+      try (InputStream is = contentSupplier.get()) {
+        final List<String> types = mimeSupport.detectMimeTypes(is, contentNameHint);
+        if (!types.isEmpty() && !types.contains(contentType)) {
+          log.debug("Discovered content type {} ", types);
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java]: ~~type~~ types

File: components/nexus-repository/src/test/java/org/sonatype/nexus/repository/storage/StorageTxImplTest.groovy [hunk 4]
```diff
@@ -55,10 +60,13 @@ extends TestSupport
   private AssetEntityAdapter assetEntityAdapter
   @Mock
   private Asset asset
-  @Mock
-  private InputStream inputStream
+
+  private InputStream inputStream = new ByteArrayInputStream('testContent'.bytes)
+  private StorageTxHooks hooks = new StorageTxHooks(emptyList())
+
+  private DefaultContentValidator defaultContentValidator = new DefaultContentValidator(new DefaultMimeSupport())
```
Comment (mrprescott) [inline components/nexus-repository/src/test/java/org/sonatype/nexus/repository/storage/StorageTxImplTest.groovy]: Should this be a mock? This is a UT for the StorageTxImpl.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  private final MimeSupport mimeSupport;
+
+  @Inject
+  public DefaultContentValidator(final MimeSupport mimeSupport) {
+    this.mimeSupport = checkNotNull(mimeSupport);
+  }
+
+  @Nonnull
+  @Override
+  public String determineContentType(boolean strictContentTypeValidation,
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java]: Could use a little UT.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java [hunk 6]
```diff
@@ -208,16 +209,31 @@ long countComponents(@Nullable String whereClause,
...
+                  String blobNameHint,
+                  InputStream inputStream,
+                  Iterable<HashAlgorithm> hashAlgorithms,
+                  @Nullable Map<String, String> headers,
+                  @Nullable String declaredContentType);
 
   /**
-   * Creates a new Blob and updates the given asset with a reference to it, hash metadata, size, and content type.
-   * The old blob, if any, will be deleted.
+   * Creates a new Blob and returns it's {@link AssetBlob}. Blobs created but not attached in a scope of a TX to any
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java]: ~~it's~~ its

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java [hunk 7]
```diff
@@ -185,21 +204,45 @@ public void registerWritePolicySelector(final WritePolicySelector writePolicySel
...
-    ODatabaseDocumentTx database = db;
-    if (database == null) {
-      database = databaseInstanceProvider.get().acquire();
+  /**
+   * Find content validator for repository format.
+   *
+   * If no format-specific validator is configured, the default is used.
+   */
+  @Nonnull
+  private ContentValidator validator() {
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: I'm torn, but it crossed my mind to pull this into a separate little service that just provides a potentially format-specific ContentValidator.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java [hunk 8]
```diff
@@ -409,26 +424,42 @@ public void deleteBucket(Bucket bucket) {
 
   @Override
   @Guarded(by = OPEN)
-  public BlobRef createBlob(final InputStream inputStream, Map<String, String> headers) {
+  public AssetBlob createBlob(final String blobNameHint,
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java]: The significance of 'blobNameHint' (it's used for content validation) really needs a mention in the Javadoc, since it could theoretically cause this method to throw an exception.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java [hunk 9]
```diff
@@ -409,26 +424,42 @@ public void deleteBucket(Bucket bucket) {
...
+  @Guarded(by = OPEN)
+  public void attachBlob(final Asset asset, final AssetBlob assetBlob)
   {
-    checkNotNull(inputStream);
-    checkNotNull(headers);
     checkNotNull(asset);
-    checkNotNull(hashAlgorithms);
-    checkNotNull(contentType);
+    checkNotNull(assetBlob);
+    checkArgument(!assetBlob.isAttached(), "Asset to attach is already attached");
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java]: How about "Blob is already attached to an asset"?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java (hunk 5 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java]: Added in 31e6c0919f2082bd72f497bd73f4c83a864c573f

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java (hunk 7 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: I would separate it as a service, but... where would you use it? Core has it here, while a plugin providing own validator can reference it just fine. Any 3rd party code/plugin that would use it? I'd say move it out once we have use for it :smile:

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java (hunk 8 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java]: Fixed in 20962ea7455f1028b2ba5f8432436c1dc67e0fe9

Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java]: Since we're storing this value as the blob name, shouldn't the parameter just be called `blobName` - the use of "hint" makes me thing this is optional or trivial, whereas as @mrprescott says it does have a significance.

Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java]: Ack, will rename it to `blobName`

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java [hunk 10]
```diff
@@ -95,7 +108,10 @@ public StorageFacetImpl(final BlobStoreManager blobStoreManager,
                           final @Named(ComponentDatabase.NAME) Provider<DatabaseInstance> databaseInstanceProvider,
                           final BucketEntityAdapter bucketEntityAdapter,
                           final ComponentEntityAdapter componentEntityAdapter,
-                          final AssetEntityAdapter assetEntityAdapter)
+                          final AssetEntityAdapter assetEntityAdapter,
+                          final ClientInfoProvider clientInfoProvider,
+                          final Map<String, ContentValidator> contentValidators,
+                          final DefaultContentValidator defaultContentValidator)
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: The default validator is available via `contentValidators.get(DefaultContentValidator.NAME)` so this parameter could be removed.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java [hunk 11]
```diff
@@ -439,27 +470,50 @@ public BlobRef setBlob(final InputStream inputStream, final Map<String, String>
...
+                         final Iterable<HashAlgorithm> hashAlgorithms,
+                         @Nullable final Map<String, String> headers,
+                         @Nullable final String declaredContentType)
+  {
+    checkNotNull(asset);
+
+    // Enforce write policy ahead, as we have asset here
+    BlobRef oldBlobRef = asset.blobRef();
+    if (oldBlobRef != null) {
+      if (writePolicySelector.select(asset, writePolicy) == WritePolicy.ALLOW_ONCE) {
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java]: These checks feel linty. What about:

    writePolicySelector.select(asset, writePolicy).checkUpdateAllowed();

Notice that the actual policy definitions already exist as javadocs in the WritePolicy enum, this would consolidate it.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java (hunk 7 above)
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: One benefit of extracting this out as a component/service is that you could then pass that into StorageFacetImpl rather than passing the map of validators plus the default validator. Encapsulating this logic also makes it easier to mock/test.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java (hunk 10 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: Same pattern is applied in ViewServlet/HttpResponseSender, to fully ensure that default is never `null` (cannot be even constructed without default sender).

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java [hunk 12]
```diff
@@ -207,15 +250,18 @@ private StorageTx openStorageTx(final ODatabaseDocumentTx db) {
     BlobStore blobStore = blobStoreManager.get(config.blobStoreName);
     return StateGuardAspect.around(
         new StorageTxImpl(
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: Looking at the long list of parameters here I'm thinking we should look at the factory pattern or even assisted-inject to help manage this.

**Comment by mrprescott** [comment]
+1 with various comments

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java (hunk 10 above)
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: Not entirely convinced by that argument ;) but if that's the case then I'd say this logic should be extracted out to a component or service to make it easier to mock/test (ditto for the logic in ViewServlet/HttpResponseSender)

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java [hunk 13]
```diff
@@ -207,15 +250,18 @@ private StorageTx openStorageTx(final ODatabaseDocumentTx db) {
...
+            txPrincipalName(),
             new BlobTx(blobStore),
-            database,
-            db != null, // userManagedDb
+            db,
+            userManagedDb,
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: Change this to `isUserManagedDB` to follow recommended naming patterns

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java [hunk 14]
```diff
@@ -208,16 +209,52 @@ long countComponents(@Nullable String whereClause,
...
+   * @param inputStream         the content to be streamed into blob store.
+   * @param hashAlgorithms      {@link HashAlgorithm}s to be applied while streaming to blob store, will be set
+   *                            in {@link Asset} attributes.
+   * @param headers             optional, custom headers for blob, if any.
+   * @param declaredContentType optional, the declared MIME type of the blob. See {@link ContentValidator}.
+   * @return {@link BlobRef} to the newly created blob. As side effect, passed in {@link Asset} is modified too, but is
+   * unsaved. Caller must ensure {@link #saveAsset(Asset)} is invoked before this TX ends.
    */
-  BlobRef createBlob(InputStream inputStream, Map<String, String> headers);
+  BlobRef setBlob(Asset asset,
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java]: This method is basically `createBlob` plus `attachBlob` with a minor precondition, so I wonder if we really need it compared to just calling those methods directly. Also it should ideally return `AssetBlob` for API consistency with `attachBlob`.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java (hunk 7 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: Extracted in a3269aec8f728a66cd722aad9a46e38ac50e78c3

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java (hunk 14 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java]: Agree, and also along the API consistency, I think this method should too throw IOEx just like createBlob is. WDYT?

Note: this method was left (almost) unchanged to not create a huge wave of changes littering over to existing plugins, hence did not change return type nor throw.

Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java]: I'm ok with this being done as a follow-up item. Concerning exceptions I agree that `setBlob` should also throw IOEx (either that or `createBlob` should be changed to throw unchecked exceptions via `Throwables.propagate`).

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java [hunk 15]
```diff
@@ -185,21 +199,29 @@ public void registerWritePolicySelector(final WritePolicySelector writePolicySel
...
 
-  private StorageTx openStorageTx(final ODatabaseDocumentTx db) {
-    ODatabaseDocumentTx database = db;
-    if (database == null) {
-      database = databaseInstanceProvider.get().acquire();
+  /**
+   * Returns the "principal name" to be used with current instance of {@link StorageTx}.
+   */
+  @Nonnull
+  private String txPrincipalName() {
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: I was confused by this, thought it was the name of the transaction - perhaps `principalName` or even `createdBy` is clearer (as that's where it will end up).

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java [hunk 16]
```diff
@@ -185,21 +199,29 @@ public void registerWritePolicySelector(final WritePolicySelector writePolicySel
...
+    ClientInfo clientInfo = clientInfoProvider.getCurrentThreadClientInfo();
+    if (clientInfo == null || clientInfo.getUserid() == null) {
+      return "system";
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: This looks like something that could go into a utility/helper method in nexus-security for dealing with clientInfo.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+                                     Supplier<InputStream> contentSupplier,
+                                     String contentName,
+                                     @Nullable String declaredContentType) throws IOException
+  {
+    checkNotNull(contentSupplier);
```
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java]: `checkNotNull(contentName);`?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java (hunk 16 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageFacetImpl.java]: We have many places coping with null `clientInfo` or null `clientInfo.userId` and those all differs by how those are used (in launcher, logs/MDC, feeds, etc). I agree we might clean that up, but again, that seems outside of the scope of this PR.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java (hunk 17 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java]: contentName is nullable on MimeSupport. It's "nice to have" as it improves detection reliability, but is not mandatory. On the other hand, we do enforce that blobs have names, checkNotNull is enforced on StorageTx blob creation methods. This is true even for master, as BlobStore enforces presence of mandatory headers.

Still, content validator API is not about blobs, and IMHO it would be more correct to mark it here as nullable instead.

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java [hunk 18]
```diff
@@ -539,9 +539,7 @@ private void createOrUpdateAssetAndContents(final StorageTx storageTx, final Buc
 
   private void attachBlob(final StorageTx storageTx, final Component component, final Asset asset, final InputStream in)
   {
-    final ImmutableMap<String, String> headers = ImmutableMap
-        .of(BlobStore.BLOB_NAME_HEADER, blobName(component), BlobStore.CREATED_BY_HEADER, "unknown");
-    storageTx.setBlob(in, headers, asset, singletonList(HashAlgorithm.SHA512), "application/zip");
+    storageTx.setBlob(asset, asset.name(), in, singletonList(HashAlgorithm.SHA512), null, "application/zip");
```
Comment (mcculls) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: Previously the blob name was `blobName(component)` which included the version (and repo) whereas now it's `asset.name()` which is just the components NuGet id without version. None of these names include the nominal NuGet extension of `.nupkg`.

We should be storing both id and version in the blob name, so it should go back to using `blobName(component)` and the implementation of `blobName` should probably be altered to create a name like "{id}-{version}.nupkg"?

Comment (cstamas) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: Ack

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java (hunk 17 above)
Comment (mcculls) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/DefaultContentValidator.java]: Agree, `@Nullable` makes sense if it's actually optional

**Comment by mcculls** [comment]
+1 with minor comments

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java (hunk 18 above)
Comment (cstamas) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: Fixed in 63967c7ec11d304ca2ef58ed986c82e9bb904f88

Comment (cstamas) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: And 6eb223aea9a179b0a6883b7ddc3332b6ad67e449

**Comment by cstamas** [comment]
Merged as 0d7df2408ff179718316345e6f331ad04beb1858

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java (hunk 14 above)
Comment (cstamas) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java]: Fixed as bb788dd16613bc56c579aea93cab5348b962f770 (was too simple for PR)
